src/Domain/DonationService.php

```diff
@@ -650,4 +641,18 @@ public function findAllCompleteForCustomerAsAPIModels(StripeCustomerId $stripeCu
 
         return array_map(fn(Donation $donation) => $donation->toFrontEndApiModel(), $donations);
     }
+
+    public function confirmDonationWithSavedPaymentMethod(Donation $donation, StripePaymentMethodId $paymentMethod): void
+    {
+        $paymentIntentId = $donation->getTransactionId();
+        Assertion::notNull($paymentIntentId);
+        $paymentIntent = $this->stripe->confirmPaymentIntent(
+            $paymentIntentId,
+            ['payment_method' => $paymentMethod->stripePaymentMethodId]
+        );
+        if ($paymentIntent->status !== 'succeeded') {
+            // @todo-regular-giving: create a new db field on Donation - e.g. payment_attempt_count and update here
+            // decide on a limit and log an error (or warning) if exceeded
```

Comment:
Yep – and maybe we want to auto cancel when the limit is reached? Although I guess releasing spare match funds isn't a high priority so long as the time windows for joining Regular Giving remain shorter than 1 month.

Reply:
Right. This comment isn't new it's just moved, but I've expanded it based on this in 856079fb13457abbf